src/v/redpanda/admin_server.cc

```diff
@@ -1748,3 +1753,85 @@ void admin_server::register_hbadger_routes() {
               [] { return ss::json::json_return_type(ss::json::json_void()); });
       });
 }
+
+void admin_server::register_transaction_routes() {
+    ss::httpd::transaction_json::get_all_transactions.set(
+      _server._routes,
+      [this](std::unique_ptr<ss::httpd::request> req)
+        -> ss::future<ss::json::json_return_type> {
+          if (!config::shard_local_cfg().enable_transactions) {
+              throw ss::httpd::bad_request_exception(
+                fmt_with_ctx(fmt::format, "Transaction are disabled"));
+          }
+
+          if (need_redirect_to_leader(model::tx_manager_ntp, _metadata_cache)) {
+              throw co_await redirect_to_leader(*req, model::tx_manager_ntp);
+          }
+
+          auto& tx_frontend = _partition_manager.local().get_tx_frontend();
+          if (!tx_frontend.local_is_initialized()) {
+              throw ss::httpd::bad_request_exception(
+                fmt::format("Can not get tx_frontend"));
+          }
+
+          auto res = co_await tx_frontend.local().get_all_transactions();
+          if (!res.has_value()) {
+              co_await throw_on_error(*req, res.error(), model::tx_manager_ntp);
+          }
+
+          using tx_info = ss::httpd::transaction_json::transaction_summary;
+          std::vector<tx_info> ans;
+          ans.reserve(res.value().size());
+
+          for (auto& tx : res.value()) {
+              if (tx.status == cluster::tm_transaction::tx_status::tombstone) {
+                  continue;
+              }
+
+              tx_info new_tx;
+
+              new_tx.transactional_id = tx.id;
+
+              ss::httpd::transaction_json::producer_identity pid;
+              pid.id = tx.pid.id;
+              pid.epoch = tx.pid.epoch;
+              new_tx.pid = pid;
+
+              new_tx.tx_seq = tx.tx_seq;
+              new_tx.etag = tx.etag;
+
+              // The motivation behind mapping killed to aborting is to make
+              // user not to think about the subtle differences between both
+              // statuses
+              if (tx.status == cluster::tm_transaction::tx_status::killed) {
+                  tx.status = cluster::tm_transaction::tx_status::aborting;
+              }
+              new_tx.status = ss::sstring(tx.get_status());
+
+              new_tx.timeout_ms = tx.get_timeout().count();
+              new_tx.staleness_ms = tx.get_staleness().count();
```

Comment:
while `get_timeout` explicitly returns milliseconds, `get_staleness` returns an opaque `lowres_clock::duration` type which shouldn't be counted on to always return milliseconds.

since the "input" here are both duration types, and the output is explicitly milliseconds, you should explicitly convert both into milliseconds to avoid the issue in which the the types used change, in which case this code might return an incorrect value.

Reply:
nit:

@VadimPlh I probably wasn't being clear enough before. What I was suggesting was that instead of changing `get_timeout / get_staleness` to return milliseconds, the duration type returned from those methods should be converted into milliseconds here at this call site.

why? it's really about code maintainability. this statement `new_tx.timeout_ms = tx.get_timeout().count();` will continue to behave as expected provided `tx.get_timeout()` returns milliseconds. but if it were changed in the future to return a generic duration type (e.g. lowres_clock::duration) or some other duration then a dev would need to be able to reliably find all the call sites to make sure the change was ok.

casting the duration here at this call site lets the type system work to our advantage!

Reply:
I have the same problem in get_transactions for partition, so I wanna fix it by another pr